### ISSUE TITLE
Issues with InChI I/O with /FixedH

### ISSUE THREAD

**Issue by baleland** [IUPAC-InChI/InChI#27]
# v1.07.0 windows binary: inchi-1.exe

# files:
#   inchi_fixedh_io.sdf
#   inchi_fixedh_io.sdf.txt.txt
#   inchi_fixedh_io.sdf.txt.log

CMD> inchi-1.exe inchi_fixedh_io.sdf /FixedH /RecMet
CMD> inchi-1.exe inchi_fixedh_io.sdf.txt /InChI2InChI

I would not have expected to see warnings about "Syntax errors" from an InChI with /FixedH  enabled

> Structure: 1 Syntax error (-2) in FIXED_H (18)
> Structure: 2 Syntax error (-2) in FIXED_H (18)
> Structure: 3 Syntax error (-2) in FIXED_H (18)
> Structure: 4 Syntax error (-2) in FIXED_H (18)
> Structure: 5 Syntax error (-2) in MOBILE_H (2)
> ...

[inchi_fixedh_io.sdf.txt.log](https://github.com/user-attachments/files/16285408/inchi_fixedh_io.sdf.txt.log)
[inchi_fixedh_io.sdf.txt](https://github.com/user-attachments/files/16285410/inchi_fixedh_io.sdf.txt)
[inchi_fixedh_io.sdf.gz](https://github.com/user-attachments/files/16285411/inchi_fixedh_io.sdf.gz)

**Comment by baleland** [comment]
There appears to be 50/5000 parse failures in our of our test validation standards using v1.07.0

The standards were generated with v1.05, do we expect more parse failures from 1.07?  Certainly (as noted above) we would _not_ expect parse failures for InChIs _generated_ by v1.07.

**Comment by djb-rwth** [comment]
Hi @baleland,
Thank you for reporting these bugs.
Issue #28 was marked as duplicate as the same bug generated both issues.
I can confirm that it has been resolved at my end.
I am posting you a pre-release version of `inchi-1.exe` for `Microsoft Windows` platform so you can test it on your mentioned `test validation standard`: [inchi-1.zip](https://github.com/user-attachments/files/16317585/inchi-1.zip)
Your feedback would be much appreciated.

**Comment by baleland** [comment]
I can do some testing with your provided pre-release version, but to run all our tests I would have expect to see a branch I could pull from to do more complete testing.  We can start with a test using your exe.

**Comment by baleland** [comment]
```
! Working in engineering mode

Finished processing 4992 structures: 0 errors, processing time 0:00:00.32`
```

Confirmed: no syntax error warning, no crashes

**Comment by djb-rwth** [comment]
Hi @baleland,
Thanks for reporting the initial test results and confirming that there are no crashes or further issues.
Since all previous Beta/RC versions have been pushed directly to `main` branch, the development `rwth` branch was inactive until yesterday.
I will push the revised code which apparently resolved both issues you reported very soon -- will let you know which branch will be used.
Thanks for your patience.

**Comment by baleland** [comment]
Am I correct in understanding that there should be NO syntax errors issued from the engineering executable you provided?  I still see some syntax errors - though these were not in the original set I provided.   I have yet to be able to run our full test suite due to the (presumably same) crash reported.

I've pulled the source from master which includes,
```
commit 00e10f234488ab4b22b743a5d8f6188d5ff4ebc0
Author: Djordje Baljozovic <121813753+djb-rwth@users.noreply.github.com>
Date:   Thu Jul 25 11:40:26 2024 +0200

    Version 1.07.1
```

I'm seeing this output from your engineering test executable,
```
Structure: 1 Syntax error (-2) in FIXED_H (18)
Structure: 2 Syntax error (-2) in FIXED_H (18)
Structure: 3 Syntax error (-2) in MOBILE_H_SP3_/s (8)
Structure: 4 Syntax error (-2) in FIXED_H (18)
Structure: 5 Syntax error (-2) in FIXED_H (18)
Structure: 6 Syntax error (-2) in FIXED_H (18)
Structure: 7 Syntax error (-2) in MOBILE_H_SP3_/s (8)
Structure: 8 Syntax error (-2) in FIXED_H (18)
Structure: 9 Syntax error (-2) in FIXED_H (18)
Structure: 10 Syntax error (-2) in MOBILE_H_SP3_/s (8)
```
[10_RECS.inchi.txt](https://github.com/user-attachments/files/16394694/10_RECS.inchi.txt)

**Comment by djb-rwth** [comment]
Hi @baleland,
This seems to be a bit confusing for me as the attached file is processed without any issues with the executable I attached.
In any case, you can now pull the image from `rwth` branch (not `main`) and test everything yourself.
Looking forward to receiving your feedback.

**Comment by baleland** [comment]
Do you see the Syntax error (-2) emitted for both of v1.07.0 and the engineering build executable?   Neither of those executables crash, but BOTH emit the Syntax errors warnings.   Isn't the point of the v1.07.1 fix to eliminate the Syntax error that (seemingly) can cause a crash?

If the syntax errors are still "expected", I'll have to dig into our tests to find out where the new crash is coming from.

**Comment by djb-rwth** [comment]
Hi @baleland,
I can now confirm that [10_RECS.inchi.txt](https://github.com/user-attachments/files/16394694/10_RECS.inchi.txt) indeed generates the above mentioned `Syntax error`s not only in `v.1.07.x` (incl. the engineering executable/development version), but in `v.1.06` as well.
I have just downloaded `v.1.05` and it reports errors too, only in different format.
I am attaching all the outputs and log files so you can see everything yourself: [10_RECS.inchi.zip](https://github.com/user-attachments/files/16428703/10_RECS.inchi.zip) .
Now, I am not sure if `10_RECS.inchi.txt` contains some strange inputs, but I would certainly like to rectify these error if they appear without any reason.

**Comment by baleland** [comment]
I can reliably get a crash from a test that does a parse of these InChIs in this specific order.  The crash happens with either a windows release or debug build.   I do not see the crash in your engineering bugfix executable, but the memory corruption may be "elsewhere" and not show up there.

Could you please try these input InChIs in a debug build of your choosing and look for any crash/exception?

The VS 2022 debugger is highlighting line 8922 in `ichiread.c` as the source of the crash.  The validation test is effectively running
the equivalent activity as your `inchi-1.exe` with options of `/InChI2InChI /FixedH /RecMet` but via a library built and linked into our own APIs - so perhaps more sensitive to memory corruption issues.

[inchi_crash_input.inchi.txt](https://github.com/user-attachments/files/16432880/inchi_crash_input.inchi.txt)

**Comment by djb-rwth** [comment]
Hi @baleland,
Please confirm that you used source files from `rwth` branch for compiling.

Here is the output that I get from `VS2022/MSVC 17.10.5` debugger:
```
InChI version 1, Software v. 1.07 (inchi-1 executable) rev. 9b6f1414ebf3+
Windows 64-bit Build (MS VS 2017 or later Debug) of Jul 31 2024 00:13:25

Opened log file 'D:/inchi_crash_input.txt.log'
Opened input file 'D:/inchi_crash_input.txt'
Opened output file 'D:/inchi_crash_input.txt.txt'
Opened problem file 'D:/inchi_crash_input.txt.prb'
The command line used:
"D:\code\dev\INCHI_EXE\bin2\Windows\x64\Debug\inchi-1.exe D:/inchi_crash_input.txt /InChI2InChI"
Generating standard InChI
Input format: InChI (plain identifier)
Output format: Plain text
Aux. info suppressed
Timeout per structure: 60000 msec
Up to 1024 atoms per structure

Tracing memory leaks (SLOW)

Finished processing 15 structures: 0 errors, processing time 0:00:00.02

Elapsed walltime: 47 msec.

D:\code\dev\INCHI_EXE\bin2\Windows\x64\Debug\inchi-1.exe (process 14104) exited with code 0.
```

I will do the same with `GCC` and post results soon.

**Comment by djb-rwth** [comment]
Hi @baleland,
Here is the output of `GDB`:
```
(gdb) run inchi_crash_input.txt -InChI2InChI
Starting program: /mnt/d/code/dev/inchi_exe/bin/linux/inchi-1 inchi_crash_input.txt -InChI2InChI
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
InChI version 1, Software v. 1.07 (inchi-1 executable)
Linux 64-bit Build (gcc 11.4.0) of Jul 30 2024 14:55:29

Opened log file 'inchi_crash_input.txt.log'
Opened input file 'inchi_crash_input.txt'
Opened output file 'inchi_crash_input.txt.txt'
Opened problem file 'inchi_crash_input.txt.prb'
The command line used:
"/mnt/d/code/dev/inchi_exe/bin/linux/inchi-1 inchi_crash_input.txt -InChI2InChI"
Generating standard InChI
Input format: InChI (plain identifier)
Output format: Plain text
Aux. info suppressed
Timeout per structure: 60000 msec
Up to 1024 atoms per structure


Finished processing 15 structures: 0 errors, processing time 0:00:00.00
[Inferior 1 (process 47022) exited normally]
```

I will try `Clang` as well, but thus far, everything looks fine.

**Comment by djb-rwth** [comment]
Hi @baleland,
`Clang` 's debugging output identical to `MSVC` 's.
Please let me know if you wish me to send you the whole `VS2022/MSVC` project in a `.zip` format so you can test everything yourself.

**Comment by baleland** [comment]
Have you tried a `valgrind` run on linux?  That can often be very enlightening.

**Comment by djb-rwth** [comment]
Hi @baleland,
Sure, here is the report:
```
valgrind -s ./inchi-1 inchi_crash_input.txt -InChI2InChI
==3722== Memcheck, a memory error detector
==3722== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3722== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==3722== Command: ./inchi-1 inchi_crash_input.txt -InChI2InChI
==3722==
InChI version 1, Software v. 1.07 (inchi-1 executable)
Linux 64-bit Build (gcc 11.4.0) of Jul 30 2024 14:55:29

Opened log file 'inchi_crash_input.txt.log'
Opened input file 'inchi_crash_input.txt'
Opened output file 'inchi_crash_input.txt.txt'
Opened problem file 'inchi_crash_input.txt.prb'
The command line used:
"./inchi-1 inchi_crash_input.txt -InChI2InChI"
Generating standard InChI
Input format: InChI (plain identifier)
Output format: Plain text
Aux. info suppressed
Timeout per structure: 60000 msec
Up to 1024 atoms per structure


Finished processing 15 structures: 0 errors, processing time 0:00:00.21
==3722==
==3722== HEAP SUMMARY:
==3722==     in use at exit: 391 bytes in 32 blocks
==3722==   total heap usage: 1,080 allocs, 1,048 frees, 4,835,034 bytes allocated
==3722==
==3722== LEAK SUMMARY:
==3722==    definitely lost: 391 bytes in 32 blocks
==3722==    indirectly lost: 0 bytes in 0 blocks
==3722==      possibly lost: 0 bytes in 0 blocks
==3722==    still reachable: 0 bytes in 0 blocks
==3722==         suppressed: 0 bytes in 0 blocks
==3722== Rerun with --leak-check=full to see details of leaked memory
==3722==
==3722== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
Please let me know if you think I might assist you with the problem you are encountering.

**Comment by baleland** [comment]
```
i 1 iComponent 0
nNum_H(iComponent + i) 1 000001C3C54A7780
nNum_H(iComponent)     1 000001C3C54A7550
pInChI[iComponent + i].nNumberOfAtoms  12
sizeof(nNum_H(0)[0])  1

i 1 iComponent 0
nNum_H(iComponent + i) 1 0000000000000000
nNum_H(iComponent)     1 000001C3C54A7640
pInChI[iComponent + i].nNumberOfAtoms  12
sizeof(nNum_H(0)[0])  1
```
is a dump of the args to memcpy() just prior to 8982 in ichiread.c
```
memcpy(nNum_H(iComponent + i), nNum_H(iComponent), pInChI[iComponent + i].nNumberOfAtoms * sizeof(nNum_H(0)[0]));
```
so `nNum_H(iComponent + i) 1`  of `0000000000000000`  will definitely be a problme

**Comment by baleland** [comment]
this looks suspicious - namely that TAUT_YES vs !TAUT_YES will bypass the alloc,
```
                       if (j == TAUT_YES && pINChI[iINChI][k][j] && pINChI[iINChI][k][j]->nNumberOfAtoms &&
                             !pINChI[iINChI][k][j]->nNum_H_fixed)
                        {
                            /* serializer crashes if it is not allocated */
                            pINChI[iINChI][k][j]->nNum_H_fixed = (S_CHAR *) inchi_calloc( (long long)pINChI[iINChI][k][j]->nNumberOfAtoms + 1, sizeof( pINChI[0][0][0]->nNum_H_fixed[0] ) ); /* djb-rwth: cast operator added */
                        }
```

**Comment by baleland** [comment]
by adding the nullptr check before the memcpy call - I can at least get our full tests to run w/o crashing, though we see more errors from the test using v1.07.1 vs the standard (v1.05 - which had NO errors, used to establish the test baseline)
```
618: Total: 5000 molecules processed.
618: Parse failures:     65
618: Parse I/O failures: 315
```
and another all-atoms-in-ptable test that fails for all 3 test structures (previously 0 failures)

something is amiss with the allocation of nNum_H_fixed, and just the "skip" to avoid the crash just means there is some lurking differences b/w v1.05 and v1.07.1 for the use of /FixedH

**Comment by baleland** [comment]
looks like the main in `e_ichimain.c` (line 690) is closer to how we are generating the InchIs in our tests, but I gather that is not the source used for inchi-1.exe

**Comment by baleland** [comment]
what is the current status of this report?  is there more information needed to reproduce in your dev environment?

**Comment by djb-rwth** [comment]
Hi @baleland,
Thank you for all the reports and sorry for a slightly delayed response.

As you might have noticed, `v.1.07.1` has been pushed/uploaded into the `main` branch. I applied some fixes in `ichiread.c` (including `NULL` pointer check you mentioned) so please test if this push has brought any changes.

I think that `pInChI[iComponent + i].nNum_H_fixed` is allocated [here](https://github.com/IUPAC-InChI/InChI/blob/f8112db2168bb0f00456685ec3fd418df84a1c3c/INCHI-1-SRC/INCHI_BASE/src/ichiread.c#L8391); a global variable named `pi_nnc3` has been added so it could be accessed within `ichiread.c` more easily.

`ichiread.c` seems to be quite different in versions `v.1.07` and `v.1.06` in comparison to `v.1.05`, as there have been some major changes throughout the code. I am only getting acquainted with `v.1.05` now, as the only reference we have all used until now was `v.1.06`.

`e_ichimain.c` belongs to `inchi_main` project, whose main purpose is to build executables which use `API` library `libinchi` (`.dll/.so/.dylib` depending on the OS). You can use this approach as well, and see if you can integrate it into your testing environment. However, please note that [several options are not available](https://github.com/IUPAC-InChI/InChI?tab=readme-ov-file#experimental-features-under-development) in the `API` version of `InChI`.

Hope this helps.

**Comment by baleland** [comment]
I'm testing with the updated source on `main`.   Indeed, some of the crashes have been fixed, but there are (now) 2 additional issues for which I've entered bugreports.

I only referred to `e_ichimain.c` to indicate the manner in which we are calling the GetINCHI() API call - namely by memset()ing all the input structs to 0.

Alll our testing was initially setup with v1.05, and refined from the v1.06 source update.  I am merely attempting to get these same tests running (and passing!) with v1.07.   Given the rather tight timeline for inclusion in our next toolkit release and the outstanding issues currently identified, it seems unlikely an InChI update will make it into our release candidates.
